contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol

```diff
@@ -0,0 +1,424 @@
+
+        if (order.takerFee != 0) {
+            if (takerFeeToken == makerToken) {
+                // Taker fee is payable in the maker token.
+                // Increase the fill amount to account for maker tokens being
+                // lost to the taker fee.
+                // makerTokenFillAmount' =
+                //  (order.makerAssetAmount * makerTokenFillAmount) /
+                //  (order.makerAssetAmount - order.takerFee)
+                makerTokenFillAmount = LibMathV06.getPartialAmountCeil(
+                    order.makerAssetAmount,
+                    order.makerAssetAmount.safeSub(order.takerFee),
+                    makerTokenFillAmount
+                );
+                // Approve the proxy to spend the maker token.
+                // It isn't worth computing the actual taker fee
+                // since `approveIfBelow()` will set the allowance to infinite. We
+                // just need a reasonable upper bound to avoid unnecessarily re-approving.
+                takerFeeToken.approveIfBelow(erc20Proxy, order.takerFee);
+            } else if (takerFeeToken != takerToken) {
+                //  Only support taker or maker asset denominated taker fees.
+                LibTransformERC20RichErrors.InvalidTakerFeeTokenError(
+                    address(takerFeeToken)
+                ).rrevert();
+            }
+        }
+
+        // Convert maker fill amount to taker fill amount.
+        uint256 takerTokenFillAmount = LibSafeMathV06.min256(
+            order.takerAssetAmount,
+            LibMathV06.getPartialAmountCeil(
+                makerTokenFillAmount,
+                order.makerAssetAmount,
+                order.takerAssetAmount
+            )
+        );
+
+        // Perform the fill.
+        return _fillOrder(
+            order,
+            signature,
+            takerTokenFillAmount,
+            protocolFee,
+            makerToken,
+            takerFeeToken == takerToken
+        );
+    }
+
+    /// @dev Attempt to fill an order. If the fill reverts, the revert will be
+    ///      swallowed and `results` will be zeroed out.
+    /// @param order The order to fill.
+    /// @param signature The order signature.
+    /// @param takerAssetFillAmount How much taker asset to fill.
+    /// @param protocolFee The protocol fee needed to fill this order.
+    /// @param makerToken The maker token.
+    /// @param isTakerFeeInTakerToken Whether the taker fee token is the same as the
+    ///        taker token.
+    function _fillOrder(
+        IExchange.Order memory order,
+        bytes memory signature,
+        uint256 takerAssetFillAmount,
+        uint256 protocolFee,
+        IERC20TokenV06 makerToken,
+        bool isTakerFeeInTakerToken
+    )
+        private
+        returns (FillOrderResults memory results)
+    {
+        // Track changes in the maker token balance.
+        results.makerTokenBoughtAmount = makerToken.balanceOf(address(this));
```

Comment:
Nit - the semantics here is sort of confusing. The way this variable is used within this function, the maker amount bought is actually zero at this point. Might just do:
```
uint256 initBalance = makerToken.balanceOf(address(this));
```